File: others/bm.java

public class BM {
    private static final int SIZE = 256;
    private void generateBC(char[] b, int m, int[] bc) {
        for (int i = 0; i < SIZE; ++i) {
            bc[i]= -1;
        }
        for (int i = 0; i < m; ++i) {
            int ascii = (int)b[i];
            bc[ascii] = i;
        }
    }

    private void generateGS(char[] b, int m, int[] suffix, boolean[] prefix) {
        for (int i = 0; i < m; ++i) {
            suffix[i] = -1;
            prefix[i] = false;
        }
        for (int i = 0; i < m - 1; ++i) {
            int j = i;
            int k = 0;
            while (j >=0 && b[j] == b[m-1-k]) {
                --j;
                ++k;
                suffix[k] = j+1;
            }
            if (j == -1) prefix[k] = true;
        }
    }

    private int moveByGS(int j, int m, int[] suffix, boolean[] prefix) {
        int k = m - 1 - j;
        if (suffix[k] != -1) return j - suffix[k] + 1;
        for (int r = j+2; r <= m-1; ++r) {
            if (prefix[m-r] == true) {
                return r;
            }
        }
        return m;
    }

    public int bm(char[] a, int n, char[] b, int m) {
        int[] bc = new int[SIZE];
        generateBC(b, m, bc);
        int[] suffix = new int[m];
        boolean[] prefix = new boolean[m];
        generateGS(b, m, suffix, prefix);
        int i = 0;
        while (i <= n-m) {
            int j;
            for (j = m -1; j >=0; --j) {
                if (a[i+j] != b[j]) break;
            }
            if (j < 0) {
                return i;
            }
            int x = j - bc[(int)a[j+i]];
            int y = 0;
            if (j < m-1) {
                y = moveByGS(j, m, suffix, prefix);
            }
            i = i + Math.max(x, y);
        }
        return -1;
    }
}

File: others/bm.py
from typing import List, Tuple

def generateBadCharacters(pattern: str) -> List[int]:
    bc = [-1 for _ in range(256)]
    for idx, char in enumerate(pattern):
        ascii = ord(char)
        bc[ascii] = idx
    return bc

def generateGoodSuffix(pattern: str) -> Tuple[List[int], List[bool]]:
    suffix = [-1 for _ in range(len(pattern))]
    prefix = [False for _ in range(len(pattern))]
    for i in range(len(pattern)-1):
        j = i
        k = 0
        while (j >= 0 and pattern[j] == pattern[len(pattern) - 1 - k]):
            k += 1
            suffix[k] = j
            j -= 1
        if j == -1: prefix[k] = True
    return suffix, prefix

def moveByGoodSuffix(bad_idx: int, pattern_len: int, \
                     suffix: List[int], prefix: List[bool]) -> int:
    k = pattern_len - 1 - bad_idx
    if suffix[k] != -1: return bad_idx + 1 - suffix[k]
    for begin in range(bad_idx+2, pattern_len):
        if prefix[pattern_len-begin] == True:
            return begin
    return pattern_len

def bm(text: str, pattern: str) -> int:
    bc = generateBadCharacters(pattern)
    suffix, prefix = generateGoodSuffix(pattern)
    i = 0
    while i <= len(text) - len(pattern):
        j = len(pattern) - 1
        while j >= 0:
            if text[i+j] != pattern[j]: break
            j -= 1
        if j < 0: return i
        x = j - bc[ord(text[i+j])]
        y = 0
        if j < len(pattern) - 1:
            y = moveByGoodSuffix(j, len(pattern), suffix, prefix)
        i = i + max(x, y)

    return -1

if __name__ == '__main__':
    print(bm('hello world', 'world'))